src/lib.rs

```diff
@@ -40,19 +42,410 @@ mod repl_helper;
+        /// NOTE: it is the responsibility of the invoking code to call machine_free() or expect
+        /// memory leaks.
+        pub extern "C" fn scryer_machine_new() -> *mut Machine {
+            let machine = Box::into_raw(Box::new(Machine::new_lib()));
+            machine // This returns a raw pointer
+        }
+
+
+        /// Frees the memory occupied by a `Machine` object.
+        ///
+        /// # Safety
+        ///
+        /// The function is marked as unsafe because it dereferences a raw pointer and assumes ownership of the object pointed to.
+        /// It relies on the `Drop` implementation of `Box` to deallocate the memory.
+        ///
+        /// # Parameters
+        ///
+        /// - `ptr`: A mutable raw pointer to a `Machine` object.
+        ///
+        /// # Example
+        ///
+        /// ```rust
+        /// use std::ptr;
+        ///
+        /// // Assume `machine_ptr` is a valid raw pointer to a `Machine` object
+        ///
+        /// unsafe {
+        ///     let machine_ptr: *mut Machine = ...;
+        ///     scryer_machine_free(machine_ptr);
+        ///     // The memory occupied by the `Machine` object is now freed
+        /// }
+        /// ```
+        #[no_mangle]
+        pub extern "C" fn scryer_machine_free(ptr: *mut Machine) {
+            unsafe { drop(Box::from_raw(ptr)); }
+        }
+
+        /// This function initiates a new query on the provided machine instance with the given 
+        /// Prolog string query as input. It uses FFI and is expected to be called from other 
+        /// languages, and thus works with raw (C) pointers.
+        /// 
+        /// the provided C string is not a valid UTF-8 string, the function panics with an error 
+        /// message. After successfully converting the C string to a Rust UTF-8 string, a new query
+        /// is started on the machine instance.
+        ///
+        /// In case of a panic during the execution, the function prints "Panic: " followed by the panic
+        /// information, and returns a null pointer. Otherwise, it returns a raw pointer to the created
+        /// QueryState.
+        ///
+        /// # Safety
+        ///
+        /// This function contains unsafe Rust code blocks, caveat emptor. The behavior is undefined if:
+        /// * `machine` is not a valid pointer to a life-alive `Machine` instance.
+        /// * `input` is not a null-terminated array.
+        ///
+        /// # Parameters
+        ///
+        /// * `machine`: Raw mutable pointer to a `Machine` instance.
+        /// * `input`:   Raw immutable pointer to a C string containing the Prolog string query.
+        ///
+        /// # Returns
+        ///
+        /// A raw pointer (`*mut QueryState`) to the created `QueryState` or a null pointer in case of a panic.
+        #[no_mangle]
+        pub extern "C" fn scryer_start_new_query_generator(machine: *mut Machine, input: *const c_char) -> *mut QueryState {
```

Comment:
It looks to me as if the caller needs to uphold the invariants of an `&mut Machine` so instead of taking a raw pointer we could be taking a mutable reference instead.

This would allow us to get rid of the unsafe block around what is currently a raw pointer dereference. We would also probably also be able to get rid of the explicit dereference al together. I think this can be applied to to all function taking `*mut Machine` except for `scryer_machine_free`.

Reply:
Though you probably still want to keep those requirements in the safety section, even though they would be implied by the type delcaration.

Reply:
But notice that functions that take `*const c_char` as argument and then turn them into `CStr` or `CString` still need to be marked `unsafe`, because there is a contract there about being a pointer to a null-terminated string.